src/util/address.rs

```diff
@@ -523,23 +531,23 @@ impl Address {
 
     /// Get the address type of the address.
     /// None if unknown, non-standard or related to the future witness version.
-    pub fn address_type(&self) -> Option<AddressType> {
+    pub fn address_type(&self) -> Result<AddressType, Error> {
         match self.payload {
-            Payload::PubkeyHash(_) => Some(AddressType::P2pkh),
-            Payload::ScriptHash(_) => Some(AddressType::P2sh),
+            Payload::PubkeyHash(_) => Ok(AddressType::P2pkh),
+            Payload::ScriptHash(_) => Ok(AddressType::P2sh),
             Payload::WitnessProgram {
                 version,
                 program: ref prog,
             } => {
                 // BIP-141 p2wpkh or p2wsh addresses.
                 match version {
                     WitnessVersion::V0 => match prog.len() {
-                        20 => Some(AddressType::P2wpkh),
-                        32 => Some(AddressType::P2wsh),
-                        _ => None,
+                        20 => Ok(AddressType::P2wpkh),
+                        32 => Ok(AddressType::P2wsh),
+                        len => Err(Error::InvalidWitnessProgramLength(len)),
                     },
-                    WitnessVersion::V1 if prog.len() == 32 => Some(AddressType::P2tr),
-                    _ => None,
+                    WitnessVersion::V1 if prog.len() == 32 => Ok(AddressType::P2tr),
+                    version => Err(Error::UnsupportedWitnessVersion(version)),
```

Comment:
nit: Can we also add len field over here? Because this will show an error as `UnsupportedWitnessVersion(1)` or `0` which might be confusing